Source/WebCore/platform/graphics/skia/FontSkia.cpp

```diff
@@ -177,4 +178,97 @@ bool Font::platformSupportsCodePoint(char32_t character, std::optional<char32_t>
     return m_platformData.skFont().getTypeface()->unicharToGlyph(character);
 }
 
+static inline float harfBuzzPositionToFloat(hb_position_t value)
+{
+    return static_cast<float>(value) / (1 << 16);
+}
+
+GlyphBufferAdvance Font::applyTransforms(GlyphBuffer& glyphBuffer, unsigned beginningGlyphIndex, unsigned beginningStringIndex, bool enableKerning, bool requiresShaping, const AtomString& locale, StringView text, TextDirection textDirection) const
+{
+    if (!enableKerning && !requiresShaping)
+        return makeGlyphBufferAdvance();
+
+    auto* hbFont = platformData().hbFont();
+    RELEASE_ASSERT(hbFont);
+
+    const auto& features = platformData().features();
+
+    bool requiresHarfbuzzShaping = [&] {
+        bool hasEnabledFeatures = features.containsIf([](const auto& feature) {
+            return !!feature.value;
+        });
+        if (hasEnabledFeatures)
+            return true;
+
+        if (enableKerning && platformData().skFont().getTypeface()->getKerningPairAdjustments({ }, { }))
+            return true;
+
+        if (!requiresShaping)
+            return false;
+
+        auto* hbFace = hb_font_get_face(hbFont);
+        if (!hbFace || !hb_ot_layout_has_substitution(hbFace))
+            return false;
+
+        return true;
+    }();
+
+    if (!requiresHarfbuzzShaping)
+        return makeGlyphBufferAdvance();
+
+    // Kerning is not handled as font features, so only in case it's explicitly disabled
+    // we need to create a new vector to include kern feature.
+    const hb_feature_t* featuresData = features.isEmpty() ? nullptr : features.span().data();
+    unsigned featuresSize = features.size();
+    Vector<hb_feature_t> featuresWithKerning;
+    if (!enableKerning) {
+        featuresWithKerning.reserveInitialCapacity(featuresSize + 1);
+        static hb_feature_t kernFeature { HB_TAG('k', 'e', 'r', 'n'), 0, 0, static_cast<unsigned>(-1) };
```

Comment:
This could be `const`, or even `constexpr`:

```suggestion
        static constexpr hb_feature_t kernFeature { HB_TAG('k', 'e', 'r', 'n'), 0, 0, static_cast<unsigned>(-1) };
```